### PR TITLE
wait and waitpid

### PR DESCRIPTION
**Pull request by sajattack**
I'm not sure how to write tests for these

### CODE REVIEW

File: src/platform/src/redox/mod.rs [hunk 1]
```diff
@@ -202,6 +202,10 @@ pub fn unlink(path: *const c_char) -> c_int {
     e(syscall::unlink(path)) as c_int
 }
 
+pub fn waitpid(pid: pid_t, stat_loc: *mut c_int, options: c_int) -> pid_t {
+    e(syscall::waitpid(pid as usize, stat_loc as &mut usize, options as usize))
```
Comment (jackpot51) [inline src/platform/src/redox/mod.rs]: You must make a temporary usize to copy stat to

File: src/platform/src/linux/mod.rs [hunk 2]
```diff
@@ -159,6 +159,10 @@ pub fn unlink(path: *const c_char) -> c_int {
     e(unsafe { syscall!(UNLINKAT, AT_FDCWD, path, 0) }) as c_int
 }
 
+pub fn waitpid(pid: pid_t, stat_loc: *mut c_int, options: c_int) -> pid_t {
+    e(unsafe { syscall!(WAIT4, pid, stat_loc, options) }) as pid_t
```
Comment (dlrobertson) [inline src/platform/src/linux/mod.rs]: This should be `syscall(WAIT4, pid, stat_loc, options, 0)`. The `wait4` syscall take 4 params. The fourth being a pointer to a `struct rusage`, which is null in the default case.

File: src/platform/src/redox/mod.rs [hunk 3]
```diff
@@ -202,6 +202,15 @@ pub fn unlink(path: *const c_char) -> c_int {
     e(syscall::unlink(path)) as c_int
 }
 
+pub fn waitpid(pid: pid_t, stat_loc: *mut c_int, options: c_int) -> pid_t {
+    unsafe {
+        let mut temp: usize = *stat_loc as usize;
```
Comment (jackpot51) [inline src/platform/src/redox/mod.rs]: stat_loc is allowed to be null, so the code should handle that.

File: src/wait/src/lib.rs [hunk 4]
```diff
@@ -11,7 +11,7 @@ use resource::rusage;
 
 #[no_mangle]
 pub unsafe extern "C" fn wait(stat_loc: *mut c_int) -> pid_t {
-    unimplemented!();
+    waitpid(0 - 1, stat_loc, 0)
```
Comment (jackpot51) [inline src/wait/src/lib.rs]: It is better to use `!0` rather than `0 - 1`, as that implies an overflow for usize.

Comment (sajattack) [reply src/wait/src/lib.rs]: Does that have the same effect? The posix spec says it should be (pid_t)-1

Comment (jackpot51) [reply src/wait/src/lib.rs]: Yes

File: tests/waitpid.c [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+int main(int argc, char** argv) {
+    pid_t pid = fork();
+    if (pid == 0) {
+    // child
+    sleep(1);
+    exit(0);
+} else {
+    // parent
+    int* stat_loc;
```
Comment (jackpot51) [inline tests/waitpid.c]: This should be int stat_loc;

File: tests/waitpid.c [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    waitpid(pid, stat_loc, 0);
```
Comment (jackpot51) [inline tests/waitpid.c]: This should use &stat_loc

File: src/platform/src/redox/mod.rs [hunk 7]
```diff
@@ -202,6 +202,17 @@ pub fn unlink(path: *const c_char) -> c_int {
     e(syscall::unlink(path)) as c_int
 }
 
+pub fn waitpid(pid: pid_t, stat_loc: *mut c_int, options: c_int) -> pid_t {
+    unsafe {
+        let mut temp: usize = 0;
+        let mut res = e(syscall::waitpid(pid as usize, &mut temp, options as usize));
```
Comment (jackpot51) [inline src/platform/src/redox/mod.rs]: res does not need to be mutable

**Comment by sajattack** [comment]
Dunno what's up with that test  fail, it works on my machine

File: tests/waitpid.c [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    if (pid == 0) {
+    // child
+    sleep(1);
+    exit(0);
+} else {
+    // parent
+    int stat_loc;
+    waitpid(pid, &stat_loc, 0);
+    }
+}
```
Comment (jackpot51) [inline tests/waitpid.c]: Returning 0 from main will fix the test failure

File: src/resource/cbindgen.toml [hunk 9]
```diff
@@ -1,7 +1,7 @@
-sys_includes = ["sys/types.h", "sys/time.h"]
+sys_includes = ["sys/types.h", "stdint.h", "sys/time.h"]
 include_guard = "_SYS_RESOURCE_H"
 language = "C"
-style = "Tag"
+style = "Both"
```
Comment (dlrobertson) [inline src/resource/cbindgen.toml]: Nice!

File: tests/waitpid.c [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    pid_t pid = fork();
+    if (pid == 0) {
+    // child
+    sleep(1);
+    exit(0);
+} else {
+    // parent
+    int stat_loc;
+    waitpid(pid, &stat_loc, 0);
+    }
```
Comment (dlrobertson) [inline tests/waitpid.c]: I agree with @MggMuggins could you fix the whitespace here?

**Pull request merged**